packages/protocol/contracts/test/locking/TimeLockManager.sol

```diff
@@ -36,6 +36,7 @@ contract TimeLockManager is ILockManager, TimeHelpers {
      * @param _start The start of the time interval
      * @param _end The end of the time interval
      */
+    // Note: not sure if there were any discussions around this, but this interaction pattern is awkward as you approve to one contract and then call another one
     function lock(IStakingLocking _staking, address _owner, uint256 _amount, uint256 _unit, uint256 _start, uint256 _end) external {
```

Comment:
I wanted to point this out, just to check if I understand how this is supposed to be used.

To lock tokens into the timelock, the user would:

1. Approve to the Staking contract
2. Call `lock()` on this TimeLockManager

Reply:
Yes, there was discussion around it (cc @facuspagnuolo ). That’s the whole point of the lock allowance. See https://github.com/aragon/staking/blob/master/packages/protocol/contracts/Staking.sol#L41 and https://github.com/aragon/staking/blob/master/packages/protocol/contracts/Staking.sol#L101 for instance.
The user likely will already have the tokens staked in the staking contract, but of course that also implies a previous interaction with it.

Reply:
Gotcha, I think I also misunderstood what `_staking.lock()` was doing at the time (it doesn't actually pull tokens from your account).

So actually, to use this lock manager (and I guess all managers, unless they've explicitly implemented a `stakeFor()`), the user would:

1. Lock tokens in the Staking contract
2. Allow this lock manager
3. Call `lock()` here

Or combine 2 & 3 with `allowManagerAndLock()`? But in this case, we wouldn't actually be registering with the logic here...?

Reply:
Yes, I would re-phrase 1 as “Stake tokens in the Staking contract”, though.
About last comment, yes, there are 2 different flows available:
- The first one, as you mentioned, would allow a contract to flexibly lock user’s token as needed (up to a “safe” amount, the allowance). Think of locking collateral any time the user perform an action.
- The second one, simpler, directly in the Staking contract, for a fixed purpose: user wants to lock a certain amount, for some reason, only once and with a fixed amount. E.g., a lockdrop.

Reply:
So I'm wondering if we should instead use an `approveAndCall()` mechanism instead to combine steps 2/3, as having two different flows for locking seems error-prone.

Would it be possible to enforce `lock()` to be only callable by the lock manager?

Reply:
How would exactly work `approveAndCall`? It would always stake and optionally lock too, depending on the `data` param?
About the second one, not sure. Would you then remove the `allowManagerAndLock` method too?

Reply:
> Would you then remove the allowManagerAndLock method too?

Yes. My line of thought on this is that locks should only be engaged from the lock manager itself, to avoid having two ways of doing this.

> How would exactly work `approveAndCall()`?

I was thinking of something like an additional arbitrary calldata argument, that if given, the Staking contract would then make a call to the provided `lockManager` with it.

Reply:
I like the idea of reducing the interface, especially around the locking methods

---

> I was thinking of something like an additional arbitrary calldata argument, that if given, the Staking contract would then make a call to the provided `lockManager` with it.

We tried this already and there was a reason why we couldn't trust the caller in the Lock Manager contract, maybe @bingen remembers a bit better